rust/agama-lib/src/storage/client/dasd.rs

```diff
@@ -66,6 +76,149 @@ impl<'a> DASDClient<'a> {
+            let job_path = self.format(&to_format).await?;
+            self.wait_for_format(job_path).await?;
+            // reprobe after calling format. TODO: check if it is needed or callbacks take into action and update it automatically
+            // also do we need to wait here for finish of format progress?
+            self.probe().await?;
+        }
+        Ok(())
+    }
+
+    async fn wait_for_format(&self, job_path: String) -> Result<(), ServiceError> {
+        let mut finished = false;
+        let jobs_client = JobsClient::new(
+            self.connection.clone(),
+            "org.opensuse.Agama.Storage1",
+            "/org/opensuse/Agama/Storage1/jobs",
+        )
+        .await?;
+        while !finished {
+            // active polling with 1 sec sleep for jobs
+            sleep(Duration::from_secs(1)).await;
+            let jobs = jobs_client.jobs().await?;
+            let job_pair = jobs
+                .iter()
+                .find(|(path, _job)| path.to_string() == job_path);
+            if let Some((_, job)) = job_pair {
+                finished = !job.running;
+            } else {
+                // job does not exist, so probably finished
+                finished = true;
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn config_set_diag(&self, config: &DASDConfig) -> Result<(), ServiceError> {
+        let pairs = self.config_pairs(config).await?;
+        let to_enable: Vec<&str> = pairs
+            .iter()
+            .filter(|(_system, config)| config.diag == Some(true))
+            .map(|(system, _config)| system.id.as_str())
+            .collect();
+        self.set_diag(&to_enable, true).await?;
+
+        let to_disable: Vec<&str> = pairs
+            .iter()
+            .filter(|(_system, config)| config.diag == Some(false))
+            .map(|(system, _config)| system.id.as_str())
+            .collect();
+        self.set_diag(&to_disable, false).await?;
+
+        if !to_enable.is_empty() || !to_disable.is_empty() {
+            // reprobe after calling format. TODO: check if it is needed or callbacks take into action and update it automatically
+            // also do we need to wait here for finish of format progress?
+            self.probe().await?;
+        }
+        Ok(())
+    }
+
+    async fn config_pairs(
+        &self,
+        config: &DASDConfig,
+    ) -> Result<Vec<(DASDDevice, DASDDeviceConfig)>, ServiceError> {
+        let devices = self.devices().await?;
+        let devices_map: HashMap<&str, DASDDevice> = devices
+            .iter()
+            .map(|d| (d.1.id.as_str(), d.1.clone()))
+            .collect();
+        config
+            .devices
+            .iter()
+            .map(|c| {
+                Ok((
+                    devices_map
+                        .get(c.channel.as_str())
+                        .ok_or(ServiceError::DASDChannelNotFound(c.channel.clone()))?
+                        .clone(),
```

Comment:
Is it possible that you find the same channel in more than one device? If it is possible, please, ignore the rest of this message :-)

If it is not possible, you can avoid that `clone` by using `remove` instead of `get` (and making the HashMap mutable, of course).